File: deces.c
#include <stdio.h>
#include <malloc.h>
#include <stdbool.h>
#include <stdlib.h>
#include<strings.h>
#include<string.h>

#define min(a,b) \
   ({ __typeof__ (a) _a = (a); \
       __typeof__ (b) _b = (b); \
     _a < _b ? _a : _b; })

//Beware of null bytes in C strings...
char *process(char *s_in, int n1,int n2,char *s_out) {
  int j=0;
  for (int i=n1;i<n2;i++) {
    char c = s_in[i];
    if ((c==0x09) || (c==0x0A) || (c==0x0C) || (c==0x0D) || (c==0x22)) s_out[j++]=' ';
    else if ((c>=0x61)&&(c<=0x7A)) s_out[j++]=c-0x20;
    else if ((c>=0x1F) && (c<0x7F)) s_out[j++]=c;
  }
  while ((j>0)&&(s_out[j-1]==' ')) j--;
  s_out [j]=0;
  for (j=0;s_out[j]==' '; j++);
  return s_out+j;
}

int find(char *s,int n1,int n2,char c) {
  int j;
  for (j=n1;(j<n2)&&(s[j]!=c);j++);
  return j;
}

int conv(char *line,int n1,int l) {
  char c = line[n1+l];
  line[n1+l]=0;
  int res = atoi(line+n1);
  line[n1+l]=c;
  return res;
}

int process_utf8(char *s,int n) {
  int j=0,flag=false;
  for (int i=0;i<=n;i++) {
    if ((s[i]>=0) || !flag) s[j++]=s[i];
    if (s[i]>=0) flag=false; else flag=true;
  }
  s[j]=0;
  return j-1;
}

void one(FILE *fp_in,FILE *fp_out) {
  char *line=NULL;
  char buf_nom[512],buf_prenom[512],insee_b[6],buf_commune_b[512],buf_pays_b[512],
    insee_d[6],buf_acte[512];
  char *nom,*prenom,sexe,*commune_b,*pays_b,*num_acte;
  int year_b,month_b,day_b,year_d,month_d,day_d;
  insee_b[5]=0;
  insee_d[5]=0;
  size_t n=0;
  while (true) {
    int nb = getline(&line,&n,fp_in);
    if (nb<0) break;
    nb=process_utf8(line,nb);
    if (nb<167) {fprintf(stderr,"Not enough chars\n");exit(-1);}
    int n1 = find(line,0,80,'*');
    nom = process(line,0,n1,buf_nom);
    int n2 = find(line,n1+1,80,'/');
    prenom=process(line,n1+1,n2,buf_prenom);
    switch (line[80]) {
    case '1' : sexe='H'; break;
    case '2' : sexe='F'; break;
    default : fprintf(stderr,"Sex error\n");exit(-1); 
    }
    year_b = conv(line,81,4);
    month_b = conv(line,85,2);
    day_b = conv(line,87,2);
    bcopy(line+89,insee_b,5);
    commune_b=process(line,94,94+30,buf_commune_b);
    pays_b=process(line,124,124+30,buf_pays_b);
    year_d = conv(line,154,4);
    month_d = conv(line,158,2);
    day_d = conv(line,160,2);
    bcopy(line+162,insee_d,5);
    num_acte=process(line,167,min(167+9,nb+1),buf_acte);
    fprintf(
	    fp_out,
	   "\"%s\",\"%s\",\"%c\",\"%d\",\"%d\",\"%d\",\"%s\",\"%s\",\"%s\",\"%d\",\"%d\",\"%d\",\"%s\",\"%s\"\n",
	   nom,prenom,sexe,year_b,month_b,day_b,insee_b,commune_b,pays_b,
	    year_d,month_d,day_d,insee_d,num_acte);
  }
}

int main(int argc,char **argv) {
  char name_in[512],name_out[512],num[512];
  FILE *fp_in,*fp_out;
  switch (argc) {
  case 2 :
    strcpy(name_out,"./deces-");
    strcat(name_out,argv[1]);
    strcat(name_out,".csv");
    fp_out = fopen(name_out,"w");
    if (fp_out==NULL) {fprintf(stderr,"Can't open out file: %s\n",name_out);exit(-1);}
    strcpy(name_in,"./deces-");
    strcat(name_in,argv[1]);
    strcat(name_in,".txt");
    fp_in = fopen(name_in,"r");
    if (fp_in==NULL) {fprintf(stderr,"Can't open in file: %s\n",name_in);exit(-1);}
    one(fp_in,fp_out);
    fclose(fp_out);
    fclose(fp_in);
    break;
  case 3:
    strcpy(name_out,"./deces-");
    strcat(name_out,argv[1]);
    strcat(name_out,"-");
    strcat(name_out,argv[2]);
    strcat(name_out,".csv");
    fp_out = fopen(name_out,"w");
    if (fp_out==NULL) {fprintf(stderr,"Can't open out file: %s\n",name_out);exit(-1);}
    int n1 = atoi(argv[1]);
    int n2 = atoi(argv[2]);
    for (int i = n1;i<=n2;i++) {
      sprintf(num,"%d",i);
      strcpy(name_in,"./deces-");
      strcat(name_in,num);
      strcat(name_in,".txt");
      fp_in = fopen(name_in,"r");
      if (fp_in==NULL) {fprintf(stderr,"Can't open in file: %s\n",name_in);exit(-1);}
      one(fp_in,fp_out);
      fclose(fp_in);
    }
    fclose(fp_out);
    break;
  default : fprintf(stderr,"Incorrect number of args\n");exit(-1);
  }
  return 0;
}

File: deces.zig
const std = @import("std");

pub fn conv(s: []u8) u32 {
    return std.fmt.parseUnsigned(u32, s, 10) catch 0;
}

pub fn process(s_in: []u8, buf: []u8) []u8 {
    var j: u32 = 0;
    for (s_in) |c| {
        if ((c == 0x09) or (c == 0x0A) or (c == 0x0C) or (c == 0x0D) or (c == 0x22)) {
            buf[j] = ' ';
            j += 1;
        } else if ((c >= 0x61) and (c <= 0x7A)) {
            buf[j] = c - 0x20;
            j += 1;
        } else if ((c >= 0x1F) and (c < 0x7F)) {
            buf[j] = c;
            j += 1;
        }
    }
    while ((j > 0) and (buf[j - 1] == ' ')) j -= 1;
    const end = j;
    j = 0;
    while ((buf[j] == ' ') and (j < end)) j += 1;
    return buf[j..end];
}

pub fn process_utf8(s: []u8) []u8 {
    var i: u32, var j: u32 = .{ 0, 0 };
    const n: usize = s.len;
    while (i < n) {
        if (s[i] < 128) {
            s[j] = s[i];
            i += 1;
            j += 1;
        } else {
            var inc: u32 = 0;
            if (((s[i] & 0b11100000) == 0b11000000) and (i < n - 1) and
                ((s[i + 1] & 0b11000000) == 0b10000000))
            {
                inc = 2;
            } else if (((s[i] & 0b11110000) == 0b11100000) and (i < n - 2) and
                ((s[i + 1] & 0b11000000) == 0b10000000) and
                ((s[i + 2] & 0b11000000) == 0b10000000))
            {
                inc = 3;
            } else if (((s[i] & 0b11111000) == 0b11110000) and (i < n - 3) and
                ((s[i + 1] & 0b11000000) == 0b10000000) and
                ((s[i + 2] & 0b11000000) == 0b10000000) and
                ((s[i + 3] & 0b11000000) == 0b10000000))
            {
                inc = 4;
            } else {
                inc = 1;
            }
            i += inc;
            s[j] = 255;
            j += 1;
        }
    }
    return s[0..j];
}

pub fn find(s: []u8, f: u8) usize {
    for (s, 0..) |c, i| {
        if (c == f) return i;
    }
    return s.len;
}

const buf_size: usize = 4096;
pub fn one(in_stream: *std.io.BufferedReader(buf_size, std.fs.File.Reader).Reader, out_stream: *std.io.BufferedWriter(buf_size, std.fs.File.Writer).Writer) !void {
    var buf: [1024]u8 = undefined;
    var buf_nom: [81]u8, var buf_prenom: [81]u8 = .{ undefined, undefined };
    var buf_commune_b: [80]u8, var buf_pays_b: [80]u8 = .{ undefined, undefined };
    var buf_acte: [80]u8 = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        const s = process_utf8(line);
        const n1 = find(s[0..80], '*');
        const nom = process(s[0..n1], &buf_nom);
        var prenom: []u8 = undefined;
        if (n1 < 79) {
            const n2 = find(s[n1 + 1 .. 80], '/');
            if (n2 == (80 - n1)) {
                prenom = process(s[n1 + 1 .. 80], &buf_prenom);
            } else {
                prenom = process(s[n1 + 1 .. @min(n1 + 1 + n2, 80)], &buf_prenom);
            }
        } else {
            prenom = "";
        }
        var sexe: u8 = undefined;
        if (s[80] == '1') sexe = 'H' else sexe = 'F';
        const year_b = conv(s[81..85]);
        const month_b = conv(s[85..87]);
        const day_b = conv(s[87..89]);
        const insee_b = s[89..94];
        const commune_b = process(s[94..124], &buf_commune_b);
        const pays_b = process(s[124..154], &buf_pays_b);
        const year_d = conv(s[154..158]);
        const month_d = conv(s[158..160]);
        const day_d = conv(s[160..162]);
        const insee_d = s[162..167];
        const num_acte = process(s[167..176], &buf_acte);
        try out_stream.print("\"{s}\",\"{s}\",\"{c}\",\"{d}\",\"{d}\",\"{d}\",\"{s}\",\"{s}\",\"{s}\",\"{d}\",\"{d}\",\"{d}\",\"{s}\",\"{s}\"\n", .{ nom, prenom, sexe, year_b, month_b, day_b, insee_b, commune_b, pays_b, year_d, month_d, day_d, insee_d, num_acte });
    }
}

pub fn main() !void {
    var buffer_in = [_]u8{0} ** 256;
    var buffer_out = [_]u8{0} ** 256;
    var args = std.process.args();
    _ = args.next();
    var i: usize = 0;
    var start: usize = 0;
    var end: usize = 0;
    while (args.next()) |arg| {
        if (i == 0) {
            start = try std.fmt.parseUnsigned(u32, arg, 10);
        } else {
            end = try std.fmt.parseUnsigned(u32, arg, 10);
        }
        i += 1;
    }
    switch (i) {
        1 => {
            const name_out = try std.fmt.bufPrint(&buffer_out, "deces-{d}.csv", .{start});
            const name_in = try std.fmt.bufPrint(&buffer_in, "deces-{d}.txt", .{start});
            var file_in = try std.fs.cwd().openFile(name_in, .{});
            var buf_reader = std.io.bufferedReaderSize(buf_size, file_in.reader());
            var in_stream = buf_reader.reader();
            var file_out = try std.fs.cwd().createFile(name_out, .{});
            var buf_writer = std.io.bufferedWriter(file_out.writer());
            var out_stream = buf_writer.writer();
            std.debug.print("{d} {d} {s} {s}\n", .{ start, end, name_in, name_out });
            one(&in_stream, &out_stream) catch std.debug.print("Error in one\n", .{});
            try buf_writer.flush();
            file_in.close();
            file_out.close();
        },
        2 => {
            const name_out = try std.fmt.bufPrint(&buffer_out, "deces-{d}-{d}.csv", .{ start, end });
            var file_out = try std.fs.cwd().createFile(name_out, .{});
            var buf_writer = std.io.bufferedWriter(file_out.writer());
            var out_stream = buf_writer.writer();
            for (start..end + 1) |y| {
                const name_in = try std.fmt.bufPrint(&buffer_in, "deces-{d}.txt", .{y});
                var file_in = try std.fs.cwd().openFile(name_in, .{});
                var buf_reader = std.io.bufferedReaderSize(buf_size, file_in.reader());
                var in_stream = buf_reader.reader();
                std.debug.print("{d} {d} {s} {s}\n", .{ start, end, name_in, name_out });
                one(&in_stream, &out_stream) catch std.debug.print("Error in one\n", .{});
                file_in.close();
            }
            try buf_writer.flush();
            file_out.close();
        },
        else => {
            std.debug.print("Incorrect number of arguments\n", .{});
        },
    }
}
